// pimatic/.js/lib/daemon.js

/*
#Daemonizer

Orginal from [node-init](https://github.com/frodwith/node-init/blob/master/init.coffee)
modified by Oliver Schneider.

Copyright (c) 2011 Paul Driver

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
 */
var daemon, fs, stream;

fs = require('fs');

daemon = require('daemon');

stream = require('logrotate-stream');

exports.printStatus = function(st) {
  if (st.pid) {
    console.log('Process running with pid %d.', st.pid);
    return process.exit(0);
  } else if (st.exists) {
    console.log('Pidfile exists, but process is dead.');
    return process.exit(1);
  } else {
    console.log('Not running.');
    return process.exit(3);
  }
};

exports.status = function(pidfile, cb) {
  if (cb == null) {
    cb = exports.printStatus;
  }
  return fs.readFile(pidfile, 'utf8', function(err, data) {
    var e, match, pid;
    if (err) {
      return cb({
        exists: err.code !== 'ENOENT'
      });
    } else if (match = /^\d+/.exec(data)) {
      pid = parseInt(match[0]);
      try {
        process.kill(pid, 0);
        return cb({
          pid: pid
        });
      } catch (error) {
        e = error;
        return cb({
          exists: true
        });
      }
    } else {
      return cb({
        exists: true
      });
    }
  });
};

exports.startSucceeded = function(pid) {
  if (pid) {
    return console.log('Process already running with pid %d.', pid);
  } else {
    return console.log('Started.');
  }
};

exports.startFailed = function(err) {
  console.log(err);
  return process.exit(1);
};

exports.start = function(arg) {
  var failure, logfile, pidfile, run, start, success;
  pidfile = arg.pidfile, logfile = arg.logfile, run = arg.run, success = arg.success, failure = arg.failure;
  success || (success = exports.startSucceeded);
  failure || (failure = exports.startFailed);
  logfile || (logfile = '/dev/null');
  start = function(err) {
    var logStream;
    if (err) {
      return failure(err);
    }
    if (process.env['PIMATIC_DAEMONIZED'] != null) {
      logStream = stream({
        file: logfile,
        size: '1m',
        keep: 3
      });
      process.stdout.writeOut = process.stdout.write;
      process.stderr.writeOut = process.stderr.write;
      process.stdout.write = function(string) {
        return logStream.write(string);
      };
      process.stderr.write = function(string) {
        return logStream.write(string);
      };
      process.logStream = logStream;
      return fs.writeFile(pidfile, process.pid, function(err) {
        if (err) {
          return failure(err);
        }
        return run();
      });
    } else {
      process.env['PIMATIC_DAEMONIZED'] = true;
      daemon.daemon(process.argv[1], process.argv.slice(2), {
        cwd: process.cwd()
      });
      return success();
    }
  };
  return exports.status(pidfile, function(st) {
    if (st.pid) {
      return success(st.pid, true);
    } else if (st.exists) {
      return fs.unlink(pidfile, start);
    } else {
      return start();
    }
  });
};

exports.stopped = function(killed) {
  if (killed) {
    console.log('Stopped.');
  } else {
    console.log('Not running.');
  }
  return process.exit(0);
};

exports.hardKiller = function() {
  return function(pid, cb) {
    var checkInterval, signals, timeout, tryKill;
    checkInterval = 1000;
    timeout = 10000;
    signals = ['TERM', 'QUIT', 'KILL'];
    tryKill = function(time) {
      var e, killed, sig;
      sig = "SIG" + signals[0];
      try {
        if (time === 0) {
          console.log("Sending " + sig + " to pimatic(" + pid + "), waiting for process exit...");
          process.kill(pid, sig);
        } else if (time >= timeout) {
          console.log("Process didn't shutdown in time, sending signal " + sig + " to pimatic(" + pid + "), waiting for process exit...");
          process.kill(pid, sig);
          if (signals.length > 1) {
            signals.shift();
          }
          time = 0;
        } else {
          process.kill(pid, 0);
        }
        return setTimeout((function() {
          return tryKill(time + checkInterval);
        }), checkInterval);
      } catch (error) {
        e = error;
        killed = e.code === 'ESRCH';
        if (!killed) {
          console.error("Couldn't kill process, error: " + e.message);
        }
        return cb(killed);
      }
    };
    return tryKill(0);
  };
};

exports.softKiller = function(timeout) {
  if (timeout == null) {
    timeout = 2000;
  }
  return function(pid, cb) {
    var sig, tryKill;
    sig = "SIGTERM";
    tryKill = function() {
      var e, first;
      try {
        process.kill(pid, sig);
        console.log("Waiting for pid " + pid);
        if (sig !== 0) {
          sig = 0;
        }
        first = false;
        return setTimeout(tryKill, timeout);
      } catch (error) {
        e = error;
        return cb(sig === 0);
      }
    };
    return tryKill();
  };
};

exports.stop = function(pidfile, cb, killer) {
  if (cb == null) {
    cb = exports.stopped;
  }
  if (killer == null) {
    killer = exports.hardKiller();
  }
  return exports.status(pidfile, function(arg) {
    var pid;
    pid = arg.pid;
    if (pid) {
      return killer(pid, function(killed) {
        return fs.unlink(pidfile, function() {
          return cb(killed);
        });
      });
    } else {
      return cb(false);
    }
  });
};

exports.simple = function(arg) {
  var command, killer, logfile, pidfile, run, start;
  pidfile = arg.pidfile, logfile = arg.logfile, command = arg.command, run = arg.run, killer = arg.killer;
  command || (command = process.argv[2]);
  killer || (killer = null);
  start = function() {
    return exports.start({
      pidfile: pidfile,
      logfile: logfile,
      run: run
    });
  };
  switch (command) {
    case 'start':
      return start();
    case 'stop':
      return exports.stop(pidfile, null, killer);
    case 'status':
      return exports.status(pidfile);
    case 'restart':
    case 'force-reload':
      return exports.stop(pidfile, start, killer);
    case 'try-restart':
      return exports.stop(pidfile, function(killed) {
        if (killed) {
          return exports.start({
            pidfile: pidfile,
            logfile: logfile,
            run: run
          });
        } else {
          console.log('Not running.');
          return process.exit(1);
        }
      });
    default:
      console.log('Command must be one of: ' + 'start|stop|status|restart|force-reload|try-restart');
      return process.exit(1);
  }
};
